sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/complexTypes.scala

```diff
@@ -79,3 +80,60 @@ case class CreateStruct(children: Seq[Expression]) extends Expression {
     InternalRow(children.map(_.eval(input)): _*)
   }
 }
+
+/**
+ * Creates a struct with the given field names and values
+ *
+ * @param children Seq(name1, val1, name2, val2, ...)
+ */
+case class CreateNamedStruct(children: Seq[Expression]) extends Expression {
+
+  private lazy val (nameExprs, valExprs) =
+    children.grouped(2).map { case Seq(name, value) => (name, value) }.toList.unzip
+
+  private lazy val names = nameExprs.map(_.asInstanceOf[Literal].value.toString)
+
+  override lazy val dataType: StructType = {
+    require(resolved, resolveFailureMessage)
+    val fields = names.zip(valExprs).map { case (name, valExpr) =>
+      StructField(name, valExpr.dataType, valExpr.nullable, Metadata.empty)
+    }
+    StructType(fields)
+  }
+
+  override def foldable: Boolean = children.forall(_.foldable)
+
+  override def nullable: Boolean = false
+
+  override def checkInputDataTypes(): TypeCheckResult = {
+    if (children.size % 2 != 0) {
+      TypeCheckResult.TypeCheckFailure("CreateNamedStruct expects an even number of arguments.")
+    } else {
+      val invalidNames = nameExprs.filterNot { case name =>
+        name match {
+          case NonNullLiteral(str, StringType) => true
+          case _ => false
+        }
+      }
+      if (invalidNames.size != 0) {
+        TypeCheckResult.TypeCheckFailure(
+          s"Non String Literal fields at odd position : ${invalidNames.mkString(",")}")
+      } else {
+        TypeCheckResult.TypeCheckSuccess
+      }
+    }
+  }
+
+  override def eval(input: InternalRow): Any = {
+    require(resolved, resolveFailureMessage)
+    InternalRow(valExprs.map(_.eval(input)): _*)
+  }
+
+  private def resolveFailureMessage(): String = {
```

Comment:
This is probably not necessary, as `childrenResolved` just part of the `resolved`, see the default implementation of `resolved`